### PR TITLE
UI Split Buttons

### PR DESCRIPTION
**Pull request by Amstutz**
Proposal for Split Buttons used in Calendar Revision, see [http://www.ilias.de/docu/goto_docu_wiki_wpage_4392_1357.html](http://www.ilias.de/docu/goto_docu_wiki_wpage_4392_1357.html). Month Split Button is introduced as special kind of split button in this proposal.

### CODE REVIEW

File: src/UI/Component/Button/Factory.php [hunk 1]
```diff
@@ -112,4 +112,29 @@ public function primary($label, $action);
...
 	public function close();
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *       The Split Button is used to group related actions if space is scarce while offering the most prominent as
+	 *       visible default.
+	 *   composition: >
+	 *       The Split Button is composed of a Standard Button offering the default action along with a dropdown containing all further options.
```
Comment (klees) [inline src/UI/Component/Button/Factory.php]: I think this could be misinterpreted easily, one might expect something akin to `<button><select>`. I'm not sure if I would consider this kind of button as "composed" at all.

Comment (klees) [reply src/UI/Component/Button/Factory.php]: BTW: It would help (me) a lot, if the lines would break after ~80 chars. If i remember correctly, having ~70 chars per line is considered to be readable.

File: src/UI/Component/Button/Factory.php [hunk 2]
```diff
@@ -112,4 +112,29 @@ public function primary($label, $action);
...
+	 *   effect: >
+	 *       Clicking the default directly fires the contained action. Clicking on the caret hinting the dropdown, opens it,
+	 *       displaying all further options and possibilities to select one of those.
```
Comment (klees) [inline src/UI/Component/Button/Factory.php:126]: Does the Split Button show a default action always, or does it somehow remember my last action, for some feasible definition of "last"?

File: src/UI/Component/Button/Factory.php [hunk 3]
```diff
@@ -112,4 +112,29 @@ public function primary($label, $action);
...
+	 *
+	 * rules:
+	 *   usage:
+	 *       1: >
+	 *           The options in the split MUST be related. They SHOULD be highly related, offering "more of the same"
```
Comment (klees) [inline src/UI/Component/Button/Factory.php]: Maybe rephrase to something like `The options in the split MUST be related. They SHOULD offer "more of the same"`. I somehow stumbled, when i read that "options must be related, but should be highly related.", couldn't make sense of it at first read.

File: src/UI/Component/Button/Split/Month.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+/* Copyright (c) 2016 Timon Amstutz <timon.amstutz@ilub.unibe.ch> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Component\Button\Split;
+
+use ILIAS\UI\Component\JavaScriptBindable;
+
+/**
+ * This describes the month split Button. Note that actions are bound to the month by using the JavaScriptBindable
```
Comment (klees) [inline src/UI/Component/Button/Split/Month.php:10]: Ok, we really need an abstraction for that action thingy.

File: src/UI/Component/Button/Split/Split.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+	/**
+	 * Get a split button like this, but action should be unavailable atm.
+	 *
+	 * The button will still have an action afterwards, this might be usefull
+	 * at some point where we want to reactivate the button client side.
+	 *
+	 * @return Split
+	 */
+	public function withUnavailableActions();
```
Comment (klees) [inline src/UI/Component/Button/Split/Split.php]: Do we really want/need this concept (activation)? I can see that we could use it at some point, but do we need it now, i.e. is there already a use case?

File: src/UI/Component/Button/Split/Standard.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * This describes a standard split button.
+ */
+interface Standard extends Split {
+	/**
+	 * Get the array containing the actions and labels of the split button
+	 *
+	 * @return	array (string|string)[]. Array containing keys as label and values as actions.
+	 */
+	public function getActionsAndLabels();
```
Comment (klees) [inline src/UI/Component/Button/Split/Standard.php:16]: No `getDefault` here?

**Review by klees — APPROVED** [review]
LGTM in general, nitpicking as usual.

File: src/UI/Component/Button/Factory.php (hunk 1 above)
Comment (Amstutz) [reply src/UI/Component/Button/Factory.php]: Thanks for the hint with the characters, changed my settings in PHP-Storms accordingly.

File: src/UI/Component/Button/Split/Month.php (hunk 4 above)
Comment (Amstutz) [reply src/UI/Component/Button/Split/Month.php:10]: Oh, yes. Desperately.

File: src/UI/Component/Button/Split/Split.php (hunk 5 above)
Comment (Amstutz) [reply src/UI/Component/Button/Split/Split.php]: So far not as I can see. I stumbled about it in your Button Implementation, thought about it for a minute - and thought - why not? But since we agreed to only offer things that are needed to keep our interfaces slim, we should kick it until asked for.

File: src/UI/Component/Button/Split/Standard.php (hunk 6 above)
Comment (Amstutz) [reply src/UI/Component/Button/Split/Standard.php:16]: Good point, kicked it out of month and promoted it to split, since available for all split button types.

Comment (klees) [reply src/UI/Component/Button/Split/Standard.php:16]: Lets duplicate [this change](https://github.com/ILIAS-eLearning/ILIAS/pull/378/commits/f978fa462735f19bb57dd18dac38949dce6fc51b) here. Consistency ftw.

**Comment by matthiaskunkel** [comment]
Jour Fixe, Dec 19, 2016: We appreciate and accept the pull request. Please allow to build split buttons from simple buttons.

**Comment by chfsx** [comment]
@Amstutz this PR currently has conflicts

**Comment by Amstutz** [comment]
Got replaced by https://github.com/ILIAS-eLearning/ILIAS/pull/546
